File: pinky_station/pinky_station/protocol/serializer.py
"""Serialize / deserialize PK protocol messages — mirrors C++ serializer."""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Optional

from .checksum import crc16
from .message_types import (
    FRAME_OVERHEAD,
    HEADER_FORMAT,
    HEADER_SIZE,
    CRC_SIZE,
    MAGIC,
    PROTOCOL_VERSION,
    MsgType,
)


# ---------------------------------------------------------------------------
# Parse result
# ---------------------------------------------------------------------------
class ParseResult(IntEnum):
    OK = 0
    INCOMPLETE = 1
    BAD_MAGIC = 2
    BAD_VERSION = 3
    BAD_HEADER_CRC = 4
    BAD_PAYLOAD_CRC = 5


@dataclass
class ParsedMessage:
    msg_type: MsgType = MsgType.ODOM
    sequence: int = 0
    payload: bytes = b""


# ---------------------------------------------------------------------------
# Serializer (builds framed messages)
# ---------------------------------------------------------------------------
class Serializer:
    def __init__(self) -> None:
        self._sequence = 0

    def frame(self, msg_type: MsgType, payload: bytes) -> bytes:
        """Build a complete framed message: header + payload + payload_crc."""
        seq = self._sequence & 0xFFFF
        self._sequence += 1

        # Build header without CRC (first 10 bytes)
        partial = struct.pack(
            "<2sBBHI",
            MAGIC,
            PROTOCOL_VERSION,
            int(msg_type),
            seq,
            len(payload),
        )
        header_crc = crc16(partial)
        header = partial + struct.pack("<H", header_crc)

        payload_crc = struct.pack("<H", crc16(payload))
        return header + payload + payload_crc

    # -- Sensor data (Robot -> PC) -----------------------------------------

    def serialize_odom(
        self, x: float, y: float, theta: float, vx: float, vth: float
    ) -> bytes:
        return self.frame(
            MsgType.ODOM, struct.pack("<5f", x, y, theta, vx, vth)
        )

    def serialize_imu(
        self,
        qw: float, qx: float, qy: float, qz: float,
        gx: float, gy: float, gz: float,
        ax: float, ay: float, az: float,
    ) -> bytes:
        return self.frame(
            MsgType.IMU,
            struct.pack("<10f", qw, qx, qy, qz, gx, gy, gz, ax, ay, az),
        )

    def serialize_lidar24(self, sectors: list[float]) -> bytes:
        assert len(sectors) == 24
        return self.frame(
            MsgType.LIDAR_24, struct.pack("<24f", *sectors)
        )

    def serialize_battery(
        self, voltage: float, percentage: float, status: int
    ) -> bytes:
        return self.frame(
            MsgType.BATTERY, struct.pack("<ffB", voltage, percentage, status)
        )

    def serialize_joint_state(
        self, pos_l: float, pos_r: float, vel_l: float, vel_r: float
    ) -> bytes:
        return self.frame(
            MsgType.JOINT_STATE, struct.pack("<4f", pos_l, pos_r, vel_l, vel_r)
        )

    def serialize_camera_frame(
        self, width: int, height: int, jpeg_data: bytes
    ) -> bytes:
        header = struct.pack("<HHI", width, height, len(jpeg_data))
        return self.frame(MsgType.CAMERA_FRAME, header + jpeg_data)

    def serialize_debug_log(
        self, severity: int, timestamp_ns: int, text: str
    ) -> bytes:
        payload = struct.pack("<BQ", severity, timestamp_ns)
        payload += text.encode("utf-8") + b"\x00"
        return self.frame(MsgType.DEBUG_LOG, payload)

    def serialize_ir_sensor(self, v0: int, v1: int, v2: int) -> bytes:
        return self.frame(
            MsgType.IR_SENSOR, struct.pack("<3H", v0, v1, v2)
        )

    def serialize_us_sensor(self, range_m: float) -> bytes:
        return self.frame(MsgType.US_SENSOR, struct.pack("<f", range_m))

    def serialize_robot_status(self, flags: int) -> bytes:
        return self.frame(MsgType.ROBOT_STATUS, struct.pack("<I", flags))

    # -- Commands (PC -> Robot) --------------------------------------------

    def serialize_cmd_vel(self, linear_x: float, angular_z: float) -> bytes:
        return self.frame(
            MsgType.CMD_VEL, struct.pack("<ff", linear_x, angular_z)
        )

    def serialize_set_led(
        self, command: int, pixel_mask: int, r: int, g: int, b: int
    ) -> bytes:
        return self.frame(
            MsgType.SET_LED, struct.pack("<5B", command, pixel_mask, r, g, b)
        )

    def serialize_set_lamp(
        self, mode: int, r: int, g: int, b: int, time_ms: int
    ) -> bytes:
        return self.frame(
            MsgType.SET_LAMP, struct.pack("<4BH", mode, r, g, b, time_ms)
        )

    def serialize_set_emotion(self, emotion_id: int) -> bytes:
        return self.frame(MsgType.SET_EMOTION, struct.pack("<B", emotion_id))

    def serialize_set_brightness(self, brightness: int) -> bytes:
        return self.frame(
            MsgType.SET_BRIGHTNESS, struct.pack("<B", brightness)
        )

    def serialize_nav_goal(
        self, x: float, y: float, theta: float
    ) -> bytes:
        return self.frame(
            MsgType.NAV_GOAL, struct.pack("<3f", x, y, theta)
        )

    def serialize_nav_cancel(self) -> bytes:
        return self.frame(MsgType.NAV_CANCEL, b"")

    def serialize_set_pose(
        self, x: float, y: float, theta: float
    ) -> bytes:
        return self.frame(
            MsgType.SET_POSE, struct.pack("<3f", x, y, theta)
        )

    def serialize_ping(self, timestamp_ns: int) -> bytes:
        return self.frame(MsgType.PING, struct.pack("<Q", timestamp_ns))

    def serialize_pong(
        self, echo_ts: int, server_ts: int
    ) -> bytes:
        return self.frame(MsgType.PONG, struct.pack("<QQ", echo_ts, server_ts))

    def serialize_ack(
        self, ack_seq: int, status: int, message: str = ""
    ) -> bytes:
        payload = struct.pack("<HB", ack_seq, status)
        payload += message.encode("utf-8")
        return self.frame(MsgType.ACK, payload)

    def serialize_map_data(
        self,
        width: int,
        height: int,
        resolution: float,
        origin_x: float,
        origin_y: float,
        origin_theta: float,
        data: bytes,
    ) -> bytes:
        header = struct.pack(
            "<IIffff", width, height, resolution,
            origin_x, origin_y, origin_theta,
        )
        return self.frame(MsgType.MAP_DATA, header + data)


# ---------------------------------------------------------------------------
# Parser
# ---------------------------------------------------------------------------
def parse_message(
    data: bytes,
) -> tuple[ParseResult, Optional[ParsedMessage], int]:
    """Parse one framed message from *data*.

    Returns (result, parsed_msg_or_None, bytes_consumed).
    """
    length = len(data)

    if length < HEADER_SIZE:
        return ParseResult.INCOMPLETE, None, 0

    if data[0:2] != MAGIC:
        return ParseResult.BAD_MAGIC, None, 1

    if data[2] != PROTOCOL_VERSION:
        return ParseResult.BAD_VERSION, None, HEADER_SIZE

    # Verify header CRC (over first 10 bytes)
    computed_hcrc = crc16(data[:10])
    (received_hcrc,) = struct.unpack_from("<H", data, 10)
    if computed_hcrc != received_hcrc:
        return ParseResult.BAD_HEADER_CRC, None, 1

    # Unpack header
    _, _, msg_type_raw, seq, payload_len, _ = struct.unpack_from(
        HEADER_FORMAT, data, 0
    )
    total_size = HEADER_SIZE + payload_len + CRC_SIZE

    if length < total_size:
        return ParseResult.INCOMPLETE, None, 0

    # Verify payload CRC
    payload_bytes = data[HEADER_SIZE : HEADER_SIZE + payload_len]
    computed_pcrc = crc16(payload_bytes)
    (received_pcrc,) = struct.unpack_from(
        "<H", data, HEADER_SIZE + payload_len
    )
    if computed_pcrc != received_pcrc:
        return ParseResult.BAD_PAYLOAD_CRC, None, total_size

    msg = ParsedMessage(
        msg_type=MsgType(msg_type_raw),
        sequence=seq,
        payload=bytes(payload_bytes),
    )
    return ParseResult.OK, msg, total_size


# ---------------------------------------------------------------------------
# Payload deserializers
# ---------------------------------------------------------------------------
def deserialize_odom(
    payload: bytes,
) -> tuple[float, float, float, float, float]:
    return struct.unpack("<5f", payload)


def deserialize_imu(payload: bytes) -> tuple:
    return struct.unpack("<10f", payload)


def deserialize_lidar24(payload: bytes) -> list[float]:
    return list(struct.unpack("<24f", payload))


def deserialize_battery(payload: bytes) -> tuple[float, float, int]:
    v, p, s = struct.unpack("<ffB", payload)
    return v, p, s


def deserialize_joint_state(
    payload: bytes,
) -> tuple[float, float, float, float]:
    return struct.unpack("<4f", payload)


def deserialize_camera_frame(
    payload: bytes,
) -> tuple[int, int, bytes]:
    w, h, size = struct.unpack_from("<HHI", payload, 0)
    jpeg = payload[8 : 8 + size]
    return w, h, jpeg


def deserialize_debug_log(payload: bytes) -> tuple[int, int, str]:
    severity = payload[0]
    ts = struct.unpack_from("<Q", payload, 1)[0]
    text = payload[9:].split(b"\x00", 1)[0].decode("utf-8")
    return severity, ts, text


def deserialize_ir_sensor(payload: bytes) -> tuple[int, int, int]:
    return struct.unpack("<3H", payload)


def deserialize_us_sensor(payload: bytes) -> float:
    return struct.unpack("<f", payload)[0]


def deserialize_robot_status(payload: bytes) -> int:
    return struct.unpack("<I", payload)[0]


def deserialize_cmd_vel(payload: bytes) -> tuple[float, float]:
    return struct.unpack("<ff", payload)


def deserialize_led_command(
    payload: bytes,
) -> tuple[int, int, int, int, int]:
    return struct.unpack("<5B", payload)


def deserialize_lamp_command(
    payload: bytes,
) -> tuple[int, int, int, int, int]:
    mode, r, g, b = struct.unpack_from("<4B", payload, 0)
    (time_ms,) = struct.unpack_from("<H", payload, 4)
    return mode, r, g, b, time_ms


def deserialize_emotion(payload: bytes) -> int:
    return payload[0]


def deserialize_brightness(payload: bytes) -> int:
    return payload[0]


def deserialize_nav_goal(payload: bytes) -> tuple[float, float, float]:
    return struct.unpack("<3f", payload)


def deserialize_set_pose(payload: bytes) -> tuple[float, float, float]:
    return struct.unpack("<3f", payload)


def deserialize_ping(payload: bytes) -> int:
    return struct.unpack("<Q", payload)[0]


def deserialize_pong(payload: bytes) -> tuple[int, int]:
    return struct.unpack("<QQ", payload)


def deserialize_ack(payload: bytes) -> tuple[int, int, str]:
    seq, status = struct.unpack_from("<HB", payload, 0)
    message = payload[3:].decode("utf-8") if len(payload) > 3 else ""
    return seq, status, message


def deserialize_map_data(
    payload: bytes,
) -> tuple[int, int, float, float, float, float, bytes]:
    w, h, res, ox, oy, ot = struct.unpack_from("<IIffff", payload, 0)
    data = payload[24:]
    return w, h, res, ox, oy, ot, data

File: pinky_core/src/protocol/serializer.cpp
#include "pinky_core/protocol/serializer.h"

#include <cstring>
#include <algorithm>

#include "pinky_core/protocol/checksum.h"

namespace pinky {

// ---------------------------------------------------------------------------
// Helper: append POD to buffer
// ---------------------------------------------------------------------------
namespace {

template <typename T>
void Append(std::vector<uint8_t>& buf, T value) {
  const auto* ptr = reinterpret_cast<const uint8_t*>(&value);
  buf.insert(buf.end(), ptr, ptr + sizeof(T));
}

template <typename T>
T ReadAt(const uint8_t* data) {
  T value;
  std::memcpy(&value, data, sizeof(T));
  return value;
}

}  // namespace

// ---------------------------------------------------------------------------
// Serializer
// ---------------------------------------------------------------------------
Serializer::Serializer() = default;

std::vector<uint8_t> Serializer::Frame(MsgType type,
                                       const std::vector<uint8_t>& payload) {
  std::vector<uint8_t> frame;
  frame.reserve(kFrameOverhead + payload.size());

  // Header fields (without CRC yet)
  MessageHeader header{};
  header.magic[0] = kMagicByte0;
  header.magic[1] = kMagicByte1;
  header.version = kProtocolVersion;
  header.msg_type = static_cast<uint8_t>(type);
  header.sequence = sequence_++;
  header.payload_length = static_cast<uint32_t>(payload.size());
  header.header_crc = 0;  // placeholder

  // Compute header CRC over first 10 bytes
  auto* raw = reinterpret_cast<const uint8_t*>(&header);
  header.header_crc = Crc16(raw, 10);

  // Write header
  raw = reinterpret_cast<const uint8_t*>(&header);
  frame.insert(frame.end(), raw, raw + kHeaderSize);

  // Write payload
  frame.insert(frame.end(), payload.begin(), payload.end());

  // Payload CRC
  uint16_t payload_crc = Crc16(payload.data(), payload.size());
  Append(frame, payload_crc);

  return frame;
}

// ---------------------------------------------------------------------------
// Serialize functions
// ---------------------------------------------------------------------------
std::vector<uint8_t> Serializer::SerializeOdom(const Odometry& odom) {
  std::vector<uint8_t> buf;
  buf.reserve(20);
  Append(buf, static_cast<float>(odom.x));
  Append(buf, static_cast<float>(odom.y));
  Append(buf, static_cast<float>(odom.theta));
  Append(buf, static_cast<float>(odom.vx));
  Append(buf, static_cast<float>(odom.vth));
  return buf;
}

std::vector<uint8_t> Serializer::SerializeImu(const ImuData& imu) {
  std::vector<uint8_t> buf;
  buf.reserve(40);
  Append(buf, static_cast<float>(imu.orientation.w));
  Append(buf, static_cast<float>(imu.orientation.x));
  Append(buf, static_cast<float>(imu.orientation.y));
  Append(buf, static_cast<float>(imu.orientation.z));
  Append(buf, static_cast<float>(imu.angular_velocity.x));
  Append(buf, static_cast<float>(imu.angular_velocity.y));
  Append(buf, static_cast<float>(imu.angular_velocity.z));
  Append(buf, static_cast<float>(imu.linear_acceleration.x));
  Append(buf, static_cast<float>(imu.linear_acceleration.y));
  Append(buf, static_cast<float>(imu.linear_acceleration.z));
  return buf;
}

std::vector<uint8_t> Serializer::SerializeLidar24(const LidarSectors& s) {
  std::vector<uint8_t> buf;
  buf.reserve(96);
  for (int i = 0; i < 24; ++i) {
    Append(buf, s.sectors[i]);
  }
  return buf;
}

std::vector<uint8_t> Serializer::SerializeBattery(const BatteryState& b) {
  std::vector<uint8_t> buf;
  buf.reserve(9);
  Append(buf, b.voltage);
  Append(buf, b.percentage);
  Append(buf, b.status);
  return buf;
}

std::vector<uint8_t> Serializer::SerializeJointState(const JointState& j) {
  std::vector<uint8_t> buf;
  buf.reserve(16);
  Append(buf, static_cast<float>(j.position[0]));
  Append(buf, static_cast<float>(j.position[1]));
  Append(buf, static_cast<float>(j.velocity[0]));
  Append(buf, static_cast<float>(j.velocity[1]));
  return buf;
}

std::vector<uint8_t> Serializer::SerializeCameraFrame(const CameraFrame& f) {
  std::vector<uint8_t> buf;
  buf.reserve(8 + f.jpeg_data.size());
  Append(buf, f.width);
  Append(buf, f.height);
  Append(buf, static_cast<uint32_t>(f.jpeg_data.size()));
  buf.insert(buf.end(), f.jpeg_data.begin(), f.jpeg_data.end());
  return buf;
}

std::vector<uint8_t> Serializer::SerializeDebugLog(const DebugLog& log) {
  std::vector<uint8_t> buf;
  buf.reserve(9 + log.text.size() + 1);
  Append(buf, log.severity);
  Append(buf, log.timestamp_ns);
  buf.insert(buf.end(), log.text.begin(), log.text.end());
  buf.push_back(0);  // null terminator
  return buf;
}

std::vector<uint8_t> Serializer::SerializeIrSensor(const IrSensor& ir) {
  std::vector<uint8_t> buf;
  buf.reserve(6);
  for (int i = 0; i < 3; ++i) {
    Append(buf, ir.values[i]);
  }
  return buf;
}

std::vector<uint8_t> Serializer::SerializeUsSensor(const UsSensor& us) {
  std::vector<uint8_t> buf;
  buf.reserve(4);
  Append(buf, us.range);
  return buf;
}

std::vector<uint8_t> Serializer::SerializeRobotStatus(const RobotStatus& s) {
  std::vector<uint8_t> buf;
  buf.reserve(4);
  Append(buf, s.flags);
  return buf;
}

std::vector<uint8_t> Serializer::SerializeCmdVel(const CmdVel& cmd) {
  std::vector<uint8_t> buf;
  buf.reserve(8);
  Append(buf, cmd.linear_x);
  Append(buf, cmd.angular_z);
  return buf;
}

std::vector<uint8_t> Serializer::SerializeLedCommand(const LedCommand& led) {
  std::vector<uint8_t> buf;
  buf.reserve(5);
  Append(buf, led.command);
  Append(buf, led.pixel_mask);
  Append(buf, led.r);
  Append(buf, led.g);
  Append(buf, led.b);
  return buf;
}

std::vector<uint8_t> Serializer::SerializeLampCommand(const LampCommand& l) {
  std::vector<uint8_t> buf;
  buf.reserve(6);
  Append(buf, l.mode);
  Append(buf, l.r);
  Append(buf, l.g);
  Append(buf, l.b);
  Append(buf, l.time_ms);
  return buf;
}

std::vector<uint8_t> Serializer::SerializeEmotion(uint8_t emotion_id) {
  return {emotion_id};
}

std::vector<uint8_t> Serializer::SerializeBrightness(uint8_t brightness) {
  return {brightness};
}

std::vector<uint8_t> Serializer::SerializeNavGoal(const NavGoal& goal) {
  std::vector<uint8_t> buf;
  buf.reserve(12);
  Append(buf, goal.x);
  Append(buf, goal.y);
  Append(buf, goal.theta);
  return buf;
}

std::vector<uint8_t> Serializer::SerializeNavCancel() {
  return {};
}

std::vector<uint8_t> Serializer::SerializeSetPose(const Pose2D& pose) {
  std::vector<uint8_t> buf;
  buf.reserve(12);
  Append(buf, pose.x);
  Append(buf, pose.y);
  Append(buf, pose.theta);
  return buf;
}

std::vector<uint8_t> Serializer::SerializePing(uint64_t timestamp_ns) {
  std::vector<uint8_t> buf;
  buf.reserve(8);
  Append(buf, timestamp_ns);
  return buf;
}

std::vector<uint8_t> Serializer::SerializePong(uint64_t echo_ts,
                                               uint64_t server_ts) {
  std::vector<uint8_t> buf;
  buf.reserve(16);
  Append(buf, echo_ts);
  Append(buf, server_ts);
  return buf;
}

std::vector<uint8_t> Serializer::SerializeAck(uint16_t ack_seq, uint8_t status,
                                              const std::string& message) {
  std::vector<uint8_t> buf;
  buf.reserve(3 + message.size());
  Append(buf, ack_seq);
  Append(buf, status);
  buf.insert(buf.end(), message.begin(), message.end());
  return buf;
}

std::vector<uint8_t> Serializer::SerializeMapData(const MapData& map) {
  std::vector<uint8_t> buf;
  buf.reserve(24 + map.data.size());
  Append(buf, map.width);
  Append(buf, map.height);
  Append(buf, map.resolution);
  Append(buf, map.origin_x);
  Append(buf, map.origin_y);
  Append(buf, map.origin_theta);
  buf.insert(buf.end(), map.data.begin(), map.data.end());
  return buf;
}

// ---------------------------------------------------------------------------
// ParseMessage
// ---------------------------------------------------------------------------
ParseResult ParseMessage(const uint8_t* data, size_t length,
                         ParsedMessage& msg, size_t& bytes_consumed) {
  bytes_consumed = 0;

  if (length < kHeaderSize) {
    return ParseResult::kIncomplete;
  }

  // Validate magic
  if (data[0] != kMagicByte0 || data[1] != kMagicByte1) {
    bytes_consumed = 1;  // skip one byte, try to resync
    return ParseResult::kBadMagic;
  }

  // Validate version
  if (data[2] != kProtocolVersion) {
    bytes_consumed = kHeaderSize;
    return ParseResult::kBadVersion;
  }

  // Validate header CRC
  uint16_t computed_hcrc = Crc16(data, 10);
  uint16_t received_hcrc = ReadAt<uint16_t>(data + 10);
  if (computed_hcrc != received_hcrc) {
    bytes_consumed = 1;
    return ParseResult::kBadHeaderCrc;
  }

  // Read header fields
  auto header = ReadAt<MessageHeader>(data);
  size_t total_size = kHeaderSize + header.payload_length + kCrcSize;

  if (length < total_size) {
    return ParseResult::kIncomplete;
  }

  // Validate payload CRC
  const uint8_t* payload_ptr = data + kHeaderSize;
  uint16_t computed_pcrc = Crc16(payload_ptr, header.payload_length);
  uint16_t received_pcrc =
      ReadAt<uint16_t>(data + kHeaderSize + header.payload_length);
  if (computed_pcrc != received_pcrc) {
    bytes_consumed = total_size;
    return ParseResult::kBadPayloadCrc;
  }

  // Success
  msg.msg_type = static_cast<MsgType>(header.msg_type);
  msg.sequence = header.sequence;
  msg.payload.assign(payload_ptr, payload_ptr + header.payload_length);
  bytes_consumed = total_size;
  return ParseResult::kOk;
}

// ---------------------------------------------------------------------------
// Deserialize functions
// ---------------------------------------------------------------------------
Odometry DeserializeOdom(const std::vector<uint8_t>& p) {
  Odometry o;
  o.x     = ReadAt<float>(p.data() + 0);
  o.y     = ReadAt<float>(p.data() + 4);
  o.theta = ReadAt<float>(p.data() + 8);
  o.vx    = ReadAt<float>(p.data() + 12);
  o.vth   = ReadAt<float>(p.data() + 16);
  return o;
}

ImuData DeserializeImu(const std::vector<uint8_t>& p) {
  ImuData d;
  d.orientation.w         = ReadAt<float>(p.data() + 0);
  d.orientation.x         = ReadAt<float>(p.data() + 4);
  d.orientation.y         = ReadAt<float>(p.data() + 8);
  d.orientation.z         = ReadAt<float>(p.data() + 12);
  d.angular_velocity.x    = ReadAt<float>(p.data() + 16);
  d.angular_velocity.y    = ReadAt<float>(p.data() + 20);
  d.angular_velocity.z    = ReadAt<float>(p.data() + 24);
  d.linear_acceleration.x = ReadAt<float>(p.data() + 28);
  d.linear_acceleration.y = ReadAt<float>(p.data() + 32);
  d.linear_acceleration.z = ReadAt<float>(p.data() + 36);
  return d;
}

LidarSectors DeserializeLidar24(const std::vector<uint8_t>& p) {
  LidarSectors s;
  for (int i = 0; i < 24; ++i) {
    s.sectors[i] = ReadAt<float>(p.data() + i * 4);
  }
  return s;
}

BatteryState DeserializeBattery(const std::vector<uint8_t>& p) {
  BatteryState b;
  b.voltage    = ReadAt<float>(p.data() + 0);
  b.percentage = ReadAt<float>(p.data() + 4);
  b.status     = ReadAt<uint8_t>(p.data() + 8);
  return b;
}

JointState DeserializeJointState(const std::vector<uint8_t>& p) {
  JointState j;
  j.position[0] = ReadAt<float>(p.data() + 0);
  j.position[1] = ReadAt<float>(p.data() + 4);
  j.velocity[0] = ReadAt<float>(p.data() + 8);
  j.velocity[1] = ReadAt<float>(p.data() + 12);
  return j;
}

CameraFrame DeserializeCameraFrame(const std::vector<uint8_t>& p) {
  CameraFrame f;
  f.width  = ReadAt<uint16_t>(p.data() + 0);
  f.height = ReadAt<uint16_t>(p.data() + 2);
  uint32_t jpeg_size = ReadAt<uint32_t>(p.data() + 4);
  f.jpeg_data.assign(p.data() + 8, p.data() + 8 + jpeg_size);
  return f;
}

DebugLog DeserializeDebugLog(const std::vector<uint8_t>& p) {
  DebugLog log;
  log.severity = ReadAt<uint8_t>(p.data() + 0);
  log.timestamp_ns = ReadAt<uint64_t>(p.data() + 1);
  // text starts at offset 9, null-terminated
  log.text = std::string(reinterpret_cast<const char*>(p.data() + 9));
  return log;
}

IrSensor DeserializeIrSensor(const std::vector<uint8_t>& p) {
  IrSensor ir;
  for (int i = 0; i < 3; ++i) {
    ir.values[i] = ReadAt<uint16_t>(p.data() + i * 2);
  }
  return ir;
}

UsSensor DeserializeUsSensor(const std::vector<uint8_t>& p) {
  UsSensor us;
  us.range = ReadAt<float>(p.data() + 0);
  return us;
}

RobotStatus DeserializeRobotStatus(const std::vector<uint8_t>& p) {
  RobotStatus s;
  s.flags = ReadAt<uint32_t>(p.data() + 0);
  return s;
}

CmdVel DeserializeCmdVel(const std::vector<uint8_t>& p) {
  CmdVel c;
  c.linear_x  = ReadAt<float>(p.data() + 0);
  c.angular_z = ReadAt<float>(p.data() + 4);
  return c;
}

LedCommand DeserializeLedCommand(const std::vector<uint8_t>& p) {
  LedCommand l;
  l.command    = p[0];
  l.pixel_mask = p[1];
  l.r = p[2];
  l.g = p[3];
  l.b = p[4];
  return l;
}

LampCommand DeserializeLampCommand(const std::vector<uint8_t>& p) {
  LampCommand l;
  l.mode = p[0];
  l.r = p[1];
  l.g = p[2];
  l.b = p[3];
  l.time_ms = ReadAt<uint16_t>(p.data() + 4);
  return l;
}

uint8_t DeserializeEmotion(const std::vector<uint8_t>& p) {
  return p[0];
}

uint8_t DeserializeBrightness(const std::vector<uint8_t>& p) {
  return p[0];
}

NavGoal DeserializeNavGoal(const std::vector<uint8_t>& p) {
  NavGoal g;
  g.x     = ReadAt<float>(p.data() + 0);
  g.y     = ReadAt<float>(p.data() + 4);
  g.theta = ReadAt<float>(p.data() + 8);
  return g;
}

Pose2D DeserializeSetPose(const std::vector<uint8_t>& p) {
  Pose2D pose;
  pose.x     = ReadAt<float>(p.data() + 0);
  pose.y     = ReadAt<float>(p.data() + 4);
  pose.theta = ReadAt<float>(p.data() + 8);
  return pose;
}

uint64_t DeserializePing(const std::vector<uint8_t>& p) {
  return ReadAt<uint64_t>(p.data());
}

PongData DeserializePong(const std::vector<uint8_t>& p) {
  PongData d;
  d.echo_ts   = ReadAt<uint64_t>(p.data() + 0);
  d.server_ts = ReadAt<uint64_t>(p.data() + 8);
  return d;
}

AckData DeserializeAck(const std::vector<uint8_t>& p) {
  AckData a;
  a.ack_seq = ReadAt<uint16_t>(p.data() + 0);
  a.status  = ReadAt<uint8_t>(p.data() + 2);
  if (p.size() > 3) {
    a.message.assign(p.begin() + 3, p.end());
  }
  return a;
}

MapData DeserializeMapData(const std::vector<uint8_t>& p) {
  MapData m;
  m.width        = ReadAt<uint32_t>(p.data() + 0);
  m.height       = ReadAt<uint32_t>(p.data() + 4);
  m.resolution   = ReadAt<float>(p.data() + 8);
  m.origin_x     = ReadAt<float>(p.data() + 12);
  m.origin_y     = ReadAt<float>(p.data() + 16);
  m.origin_theta = ReadAt<float>(p.data() + 20);
  if (p.size() > 24) {
    m.data.assign(p.begin() + 24, p.end());
  }
  return m;
}

}  // namespace pinky
